Extensions/TileMap/collision/TransformedTileMap.ts

```diff
@@ -701,22 +733,110 @@ namespace gdjs {
         const definition = this.getDefinition();
         if (!definition) {
           this._setHitboxesUpToDate();
-          // It should already be []
           this.hitBoxes.length = 0;
           return this.hitBoxes;
         }
         const tag = this.layer.tileMap.tag;
         const definitionHitboxes = definition.getHitBoxes(tag);
         if (!definitionHitboxes) {
           this._setHitboxesUpToDate();
-          // It should already be []
           this.hitBoxes.length = 0;
           return this.hitBoxes;
         }
 
-        const layerTransformation = this.layer.tileMap.getTransformation();
-        const width = this.layer.tileMap.getTileWidth();
-        const height = this.layer.tileMap.getTileHeight();
+        const tileMap = this.layer.tileMap;
+        const width = tileMap.getTileWidth();
+        const height = tileMap.getTileHeight();
+
+        // Extend the hit boxes.
+        // It avoids small objects to be pushed side way into a wall when they
+        // should be pop out of the wall.
+
+        const hasFullHitBox =
+          definitionHitboxes.length === 1 && definition.hasFullHitBox(tag);
+        if (hasFullHitBox) {
+          const isLeftFull = this._hasNeighborFullHitBox(-1, 0);
+          const isRightFull = this._hasNeighborFullHitBox(1, 0);
+          const isTopFull = this._hasNeighborFullHitBox(0, -1);
+          const isBottomFull = this._hasNeighborFullHitBox(0, 1);
+
+          let hitBoxesCount = 0;
+          if (isLeftFull || isRightFull) {
+            let minX = isLeftFull ? -width : 0;
+            let maxX = isRightFull ? 2 * width : width;
+            if (this.hitBoxes.length === 0) {
+              // TODO Avoid to allocate a Polygon
+              this.hitBoxes[0] = gdjs.Polygon.createRectangle(0, 0);
+            }
+            TransformedCollisionTile.setRectangle(
+              this.hitBoxes[0],
+              minX,
+              0,
+              maxX,
+              height
+            );
+            hitBoxesCount++;
+          }
+          if (isTopFull || isBottomFull) {
+            let minY = isTopFull ? -height : 0;
+            let maxY = isBottomFull ? 2 * height : height;
+            if (this.hitBoxes.length < hitBoxesCount + 1) {
+              // TODO Avoid to allocate a Polygon
+              this.hitBoxes[hitBoxesCount] = gdjs.Polygon.createRectangle(0, 0);
+            }
+            TransformedCollisionTile.setRectangle(
+              this.hitBoxes[hitBoxesCount],
+              0,
+              minY,
+              width,
+              maxY
+            );
+            hitBoxesCount++;
+          }
+          if (hitBoxesCount === 0) {
+            if (this.hitBoxes.length === 0) {
+              // TODO Avoid to allocate a Polygon
```

Comment:
Are these TODO really doable? Do we really want to keep them?

Reply:
We could recycle them, but it will only happen after a tile is changed so it's probably not worth it. I can remove the todo.